Support dbcreator fixed server role

anju15bharti opened the pull request:
**Issue Resolved**
BABEL-5115

**Description of the changes:**
Added support for new fixed server role dbcreator

This role has following privileges with it:
Members of the dbcreator fixed server role can 
1. Create database 
2. Alter database (if has access to that database, i.e either owner of the database or has mapped user in it) 
3. Drop any database. 

Engine PR : https://github.com/amazon-aurora/postgresql_modified_for_babelfish/pull/93

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Commit:
Support Securityadmin fixed server role

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Addressed Comments

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

anju15bharti committed:
Merge branch 'BABEL_4_X_DEV' into jira-babel-5040

Commit:
Updated expected_dependency file

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Added securityadmin membership check in is_babelfish_role

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/BABEL_4_X_DEV' into jira-babel-5040

Commit:
rerun test

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
ignore logs for dump & restore

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Added more testcases

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

anju15bharti committed:
Merge branch 'BABEL_4_X_DEV' into jira-babel-5040

Commit:
Fix test failure

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Revert "ignore logs for dump & restore"

This reverts commit c584f245804f1b7017eec64ca20cf2da619a2e4d.

Commit:
Support dbcreator fixed server role

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Added testfile in upgrade script

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tds/src/backend/tds/tdsutils.c:
```diff
@@ -1234,18 +1236,21 @@ handle_grant_role(GrantRoleStmt *grant_stmt)
 
 		roleid = get_role_oid(rolename, false);
-		if (OidIsValid(roleid) && (roleid == bbf_role_admin_oid || roleid == securityadmin_oid))
+		if (OidIsValid(roleid) && (roleid == bbf_role_admin_oid || roleid == securityadmin_oid
+							|| roleid == dbcreator_oid))
```
We will need to refactor this. Given that we will be introducing around 5-6 roles in the future.

tanscorpio7 replied:
Curious why sysadmin is not inlcuded here ?

HarshLunagariya commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -4882,8 +4883,11 @@ rename_tsql_db(char *old_db_name, char *new_db_name)
+
 	/* Check permission on the given database. */
-	if (!has_privs_of_role(GetSessionUserId(), get_role_oid("sysadmin", false)))
+	if (!has_privs_of_role(GetSessionUserId(), get_sysadmin_oid())
+		&& !(user && has_privs_of_role(GetSessionUserId(), get_dbcreator_oid())))
```
Can you please explain this?

anju15bharti replied:
For a login which is member of dbcreator want to alter database name then it should have access to that database(which it can have if there is a mapped user for that login in that database). Will add comment as well above the code.

tanscorpio7 replied:
Do we have a test case where dbcreator has a user but it is disabled ?

anju15bharti replied:
We can only disable login not user but yes to check this scenario by revoking connect from user for that db which will basically mean that there is no user for that db leading to fall into this case and will get the expected error. Will add in the testfile as well.

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3746,6 +3751,17 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+						if (grant_role->is_grant)
+							appendStringInfo(&query, "ALTER ROLE dummy WITH createdb; ");
+						
+						/* If grantee role is member of sysadmin then don't revoke createdb */
```
misleading comment

Commit:
Addressed comments

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Merge remote-tracking branch 'upstream/server-database-roles-4-x' into jira-babel-5040

anju15bharti committed:
Merge branch 'server-database-roles-4-x' into jira-babel-5040

Commit:
Retrigger Tests

Commit:
Fixed test failure

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Added testfile to upgrade script & handled alter login(sysadmin member) case

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Merge remote-tracking branch 'origin/server-database-roles-4-x' into jira-babel-5040

Commit:
Fix test failure

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

tanscorpio7 commented on contrib/babelfishpg_tds/src/backend/tds/tdsutils.c:
```diff
@@ -1208,12 +1208,14 @@ handle_grant_role(GrantRoleStmt *grant_stmt)
 
 	bbf_role_admin_oid = get_role_oid(BABELFISH_ROLE_ADMIN, false);
 	securityadmin_oid = get_role_oid(BABELFISH_SECURITYADMIN, false);
+	dbcreator_oid = get_role_oid(BABELFISH_DBCREATOR, false);
```
Can we cache these OIDs in this file.

tanscorpio7 commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -287,12 +287,15 @@ BEGIN
 	EXECUTE format('CREATE ROLE securityadmin CREATEROLE INHERIT PASSWORD NULL');
 	EXECUTE format('GRANT CONNECT ON DATABASE %s TO securityadmin WITH GRANT OPTION', CURRENT_DATABASE());
+	EXECUTE format('CREATE ROLE dbcreator CREATEDB INHERIT PASSWORD NULL');
+	EXECUTE format('GRANT CONNECT ON DATABASE %s TO dbcreator WITH GRANT OPTION', CURRENT_DATABASE());
```
I think we discussed this in `Securityadmin` PR that we do not need this.

tanscorpio7 replied:
Also do we need to give `dbcreator` any permission from PG perspective ? Members of this role are again granted this attribute and this role will never be the session user. Same goes for securityadmin.

anju15bharti replied:
yes the rebased commit will have those changes from securityadmin PR.

anju15bharti replied:
regarding any need giving dbcreator any permission from PG perspective, In that way we also don't need to give any permission to sysadmin but we decided to give it. So following that dbcreator and securityadmin role was given the required attribute.

tanscorpio7 replied:
So just for consistency.

tanscorpio7 commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.3.0--4.4.0.sql:
```diff
@@ -111,6 +111,25 @@ AND bbf_is_member_of_role_nosuper(sys.suser_id(), sys.suser_id('securityadmin'))
+    END IF;
+END;
+$$;
+
```
This will be a common function for securityadmin and db_creator right ?

anju15bharti replied:
yes

tanscorpio7 commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -1981,7 +1981,8 @@ is_server_role_supported (const char *serverrole, int position, core_yyscan_t yy
 	}
-	else if (strcmp(serverrole, BABELFISH_SYSADMIN) != 0 && strcmp(serverrole, BABELFISH_SECURITYADMIN) != 0)
+	else if (strcmp(serverrole, BABELFISH_SYSADMIN) != 0 && strcmp(serverrole, BABELFISH_SECURITYADMIN) != 0
+							&& strcmp(serverrole, BABELFISH_DBCREATOR) != 0)
```
Lets create a MACRO for this.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -4882,8 +4883,11 @@ rename_tsql_db(char *old_db_name, char *new_db_name)
 			(errcode(ERRCODE_OBJECT_IN_USE),
 				errmsg("The database could not be exclusively locked to perform the operation.")));
 
+	user = get_user_for_database(old_db_name);
```
Should we put it inside the if condition, only compute this if not member of sysadmin. Not a performance critical path but good to have.

anju15bharti replied:
Right since we are using this function once we can include it itself in the below if condition after sysadmin which will avoid extra check.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3732,12 +3733,16 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+						
+						/* If grantee role is member of dbcreator then only revoke createrole */
+						else if (has_privs_of_role(grantee_oid, get_dbcreator_oid()))
+							appendStringInfo(&query, "ALTER ROLE dummy WITH nocreaterole; ");
```
I think this entire code block should be refactored into another function.

We first figure out through how many server role membership is a particular attribute granted to current role and only revoke that attribute if count is one.

anju15bharti replied:
We can separate this out into another function to make it more readable, but maintaining counter to figure out which attribute to revoke will actually not be that useful here because We still need to separately grant and revoke priv from role for sysadmin, securityadmin and dbcreator because in general we cannot say through which membership we got this attribute. For example if the command it to drop member from sysadmin and we got createdb attribut via dbcreator then although the counter for createdb will be 1 yet it is not supposed to be revoked.

Still improvised the function using counter.

tanscorpio7 replied:
I was suggesting something similar to what you have already done for sysadmin.

Could you also move the increment counter conditions outside `if (IS_ROLENAME_SYSADMIN(rolename))`  this will allow all our revoke condition later on to look like
```else if (createdb == 1)``` only then revoke

anju15bharti replied:
I don't think that would be a good idea because then irrespective of if the priv check is required for particular rolename or not we will always do the priv check for both securityadmin and dbcreator( which is actually not required) i.e in securityadmin case we need not to check for dbcreator priv and in dbcreator we need not to check for securityadmin priv.

So I think keeping the way it is would be more efficient.

tanscorpio7 replied:
makes sense, then I think we can remove the entire counter approach. Just doing it for sysadmin does not really work. 
We can think of something else later on to refactor this. May be completely remove the attribute grant/revoke and just rely on memberships checks everywhere.

anju15bharti replied:
No I think we can keep it for now since this counter is helping improving the checks call for sysadmin for every role case otherwise we will have to wrtie sysadmin priv check in each case. Better than just doing every priv check separately.

anju15bharti committed:
Support Securityadmin fixed server role (#67)

Added support for new fixed server role securityadmin. This role has following privileges with it:

Members of the securityadmin fixed server role can manage logins and their properties.
They can GRANT, DENY, and REVOKE server-level permissions.
securityadmin can also GRANT, DENY, and REVOKE database-level permissions if they have access to a database.

Original PR with all the comments babelfish-for-postgresql#2907

Issue resolved: BABEL-5040
Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Merge branch 'jira-babel-5040' into jira-dbcreator

Commit:
Address comments

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

tanscorpio7 committed:
Support fixed database role db_accessadmin (#66)

* support db_accessadmin

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* empty commit to rerun actions

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* add basic tests

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix get_authid_user_ext_physical_name

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* refactor

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* refactor is_role() & is_user() to one function

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* address comments

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* empty commit to rerun actions

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* add test to upgrade

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* add test to upgrade

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix tests

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* empty

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* see dump restore actions logs

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* ssh into github actions host

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* remove temp changes and reset password for login in verify file

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* empty

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* refactor

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix tests

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* refactor

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix is_member

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix is_member

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* refactor upgrade function for creating roles

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* refactor

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix merge issue

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* refactor name for upgrade function

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* use IS_BBF_BUILT_IN_DB macro in some more places

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* empty

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix upgrade file

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

* fix typo

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

---------

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

Commit:
Merge remote-tracking branch 'origin/server-database-roles-4-x' into jira-dbcreator

Commit:
Added testfile to all 15 and 16 upgrade schedule

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Fix failure

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
rerun

Commit:
fix failure

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Merge remote-tracking branch 'origin/server-database-roles-4-x' into jira-dbcreator

Commit:
Fix upgrade failure

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

tanscorpio7 commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -1731,7 +1744,8 @@ check_alter_server_stmt(GrantRoleStmt *stmt)
 	 */
 	if (!has_privs_of_role(GetSessionUserId(), sysadmin) && ((strcmp(granted_name, BABELFISH_SECURITYADMIN) != 0)
-										|| !has_privs_of_role(GetSessionUserId(), get_securityadmin_oid())))
+		|| !has_privs_of_role(GetSessionUserId(), get_securityadmin_oid())) && ((strcmp(granted_name, BABELFISH_DBCREATOR) != 0)
+						|| !has_privs_of_role(GetSessionUserId(), get_dbcreator_oid())))
```
Can we reuse `IS_ROLENAME_SECURITYADMIN` and other macros here ? Also some other places we can reuse these macros.

Deepesh125 replied:
update comment also

tanscorpio7 commented on contrib/babelfishpg_tsql/src/dbcmds.c:
```diff
@@ -543,7 +543,10 @@ create_bbf_db_internal(ParseState *pstate, const char *dbname, List *options, co
 	{
 		SetUserIdAndSecContext(GetSessionUserId(), save_sec_context | SECURITY_LOCAL_USERID_CHANGE);
-		if (!have_createdb_privilege())
+
+		/* Session user must be member of sysadmin or dbcreator fixed server role */
+		if (!have_createdb_privilege() && !is_member_of_role(GetSessionUserId(), get_sysadmin_oid()) &&
+							!is_member_of_role(GetSessionUserId(), get_dbcreator_oid()))
```
Is this needed ? logins would already have the createdb privilege right ?

anju15bharti replied:
We should keep it as earlier we were making sure that it's sysadmin by checking if it can set sysadmin role, so better to keep this check. Also There can be possibility that login(malicious) is not member of any of the roles but has creatdb priv somehow from pg endpoint.

tanscorpio7 replied:
can we use `has_privs_of_role` instead of `is_member_of_role` since that is what we used everywhere else.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -4868,6 +4868,14 @@ rename_tsql_db(char *old_db_name, char *new_db_name)
 			(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
 				errmsg("Cannot change the name of the system database %s.", old_db_name)));
 
+	/* Check permission on the given database. */
```
Could we update the comment here to show db_creator can only alter the databases in which it has an active user.

HarshLunagariya replied:
+1, instead of "active" , let's say "mapped" user.

tanscorpio7 commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.3.0--4.4.0.sql:
```diff
@@ -197,7 +202,9 @@ BEGIN
-    ELSIF role = 'sysadmin' COLLATE sys.database_default OR role = 'securityadmin' COLLATE sys.database_default THEN
+    ELSIF role = 'sysadmin' COLLATE sys.database_default 
+		OR role = 'securityadmin' COLLATE sys.database_default 
+		OR role = 'dbcreator' COLLATE sys.database_default THEN
```
role IN ( .., .., ..)

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:
```diff
@@ -2060,7 +2060,9 @@ BEGIN
- 	ELSIF role = 'sysadmin' COLLATE sys.database_default OR role = 'securityadmin' COLLATE sys.database_default THEN
+ 	ELSIF role = 'sysadmin' COLLATE sys.database_default 
+		OR role = 'securityadmin' COLLATE sys.database_default 
+		OR role = 'dbcreator' COLLATE sys.database_default THEN
```
Refactor it to `role in (r1, r2, r2);`

HarshLunagariya commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -149,6 +150,8 @@ create_bbf_authid_login_ext(CreateRoleStmt *stmt)
 		new_record_login_ext[LOGIN_EXT_TYPE] = CStringGetTextDatum("Z");
 	else if (strcmp(stmt->role, BABELFISH_SECURITYADMIN) == 0)
 		new_record_login_ext[LOGIN_EXT_TYPE] = CStringGetTextDatum("R");
+	else if (strcmp(stmt->role, BABELFISH_DBCREATOR) == 0)
+		new_record_login_ext[LOGIN_EXT_TYPE] = CStringGetTextDatum("R");
```
Can we use macro IS_BBF_FIXED_SERVER_ROLE here to combine 2 else if condition for secadmin and dbcreator?

Deepesh125 replied:
+1

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6932,3 +6904,78 @@ bbf_set_tran_isolation(char *new_isolation_level_str)
+ * privileges to/from the login on the basis of
+ * that login's privilege.
+ */
+static void 
```
We refactor it in following way for better readability:
```
static void 
gen_command_grant_revoke_priv_to_role(StringInfo query, const char *rolename,
							bool is_grant, Oid login_oid)
{
    int createrole = 0, createdb = 0;

	if (has_privs_of_role(login_oid, get_sysadmin_oid()))
	{
		createrole++;
		createdb++;
	}
	if (has_privs_of_role(login_oid, get_securityadmin_oid()))
	{
		createrole++;
	}
	if (has_privs_of_role(login_oid, get_dbcreator_oid()))
	{
		createdb++;
	}

	/* If sysadmin, provide attribute for role and database priv */
	if (IS_ROLENAME_SYSADMIN(rolename))
	{
		createrole += ((is_grant) ? 1 : -1);
		createdb += ((is_grant) ? 1 : -1);
	}
	else if (IS_ROLENAME_SECURITYADMIN(rolename))
	{
		createrole += ((is_grant) ? 1 : -1);
	}
	else if (IS_ROLENAME_DBCREATOR(rolename))
	{
		createdb += ((is_grant) ? 1 : -1);
	}
	else
	{
		....
	}
	
	appendStringInfo(query, "ALTER ROLE dummy WITH %s %s; ", (createdb > 0) ? "createdb" : "nocreatedb", (createrole > 0) ? "createrole" : "nocreaterole");
}
```

anju15bharti replied:
But then it will unnecessarily call the priv check for securityadmin in case of dbcreator and priv check for dbcreator in case of securityadmin which doesn't seem efficient. Does it?

HarshLunagariya replied:
The 2nd and 3rd execution of `has_privs_of_role()`(internally uses `roles_is_member_of`)  will be quick, because internally it caches the member's list (based on the given role's oid) after 1st execution. So, efficiency shouldn't be issue here.

HarshLunagariya commented on test/JDBC/expected/dbcreator_role-vu-verify.out:
```diff
@@ -0,0 +1,1184 @@
+1
+~~END~~
+
+
```
```
select is_srvrolemember('sysadmin', 'dbcreator')
select is_srvrolemember('securityadmin', 'dbcreator')
And vice versa tests.
```

HarshLunagariya replied:
+ Please add testcases for sp_helpsrvrole and sp_helpsrvrolemember(if supported)

anju15bharti replied:
sp_helpsrvrole is unsupported and added test for sp_helpsrvrolemember.

HarshLunagariya commented on test/JDBC/expected/dbcreator_role-vu-verify.out:
```diff
@@ -0,0 +1,1184 @@
+use dbcreator_db1
+go
+
+grant connect to guest
+go
+
+-- terminate-tsql-conn
+
+-- tsql user=dbcreator_login1 password=123
+-- should be allowed
+EXEC sp_renamedb 'dbcreator_db1', 'dbcreator_db2'
+GO
```
Just double-checking, This is verified ideal T-SQL behaviour, right?

anju15bharti replied:
yes

Deepesh125 commented on contrib/babelfishpg_tds/src/backend/tds/tdsutils.c:
```diff
@@ -939,17 +947,20 @@ is_babelfish_role(const char *role)
 	role_oid = get_role_oid(role, true);	/* missing OK */
-	securityadmin_oid = get_role_oid(BABELFISH_SECURITYADMIN, true);  /* missing OK */
+	securityadmin = get_role_oid(BABELFISH_SECURITYADMIN, true);  /* missing OK */
+	dbcreator = get_role_oid(BABELFISH_DBCREATOR, true);  /* missing OK */
```
why missing_ok = true? we are not validating dbcreator before checking membership on line 963

tanscorpio7 replied:
I plan on completely refactoring this code area as part of BABEL-5332

Deepesh125 commented on contrib/babelfishpg_tsql/src/catalog.h:
```diff
@@ -336,6 +337,16 @@ typedef FormData_bbf_function_ext *Form_bbf_function_ext;
+	(strlen(rolname) == 9 && \
+	strncmp(rolname, BABELFISH_DBCREATOR, 9) == 0)
+
+#define IS_BBF_FIXED_SERVER_ROLE(rolename) \
+	(IS_ROLENAME_SYSADMIN(rolename) || \
+	IS_ROLENAME_SECURITYADMIN(rolename) || \
+	IS_ROLENAME_DBCREATOR(rolename))
```
its really hard to understand distinction b/w what is defined here vs IS_DEFAULT_BBF_SERVER_ROLE defined at tds_int.h

anju15bharti replied:
`IS_DEFAULT_BBF_SERVER_ROLE` includes our internal bbf role bbf_role_admin as well while other one is just for our supported fixed server role.

Deepesh125 commented on test/JDBC/expected/dbcreator_role-vu-verify.out:
```diff
@@ -0,0 +1,1184 @@
+use dbcreator_db
+go
+
+select current_user
+go
+~~START~~
+varchar
+dbo
+~~END~~
```
this is expected?

anju15bharti replied:
yes, since this login has created the database `dbcreator_db`

Commit:
Address comments

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

tanscorpio7 commented on contrib/babelfishpg_tsql/src/dbcmds.c:
```diff
@@ -600,6 +599,7 @@ create_bbf_db_internal(ParseState *pstate, const char *dbname, List *options, co
 	old_dbname = get_cur_db_name();
 	set_cur_db(dbid, dbname);	/* temporarily set current dbid as the new id */
 	dbo_role = get_dbo_role_name(dbname);
+	datdba = get_role_oid("sysadmin", false);
```
lets use `get_sysadmin_oid`

tanscorpio7 commented on contrib/babelfishpg_tsql/src/dbcmds.c:
```diff
@@ -741,7 +741,9 @@ drop_bbf_db(const char *dbname, bool missing_ok, bool force_drop)
 
-		if (!(has_privs_of_role(roleid, get_role_oid("sysadmin", false)) || login_is_db_owner))
+		/* Check if login has required privilege to drop the database */
+		if (!(has_privs_of_role(roleid, get_sysadmin_oid()) 
+			|| has_privs_of_role(roleid, get_dbcreator_oid()) || login_is_db_owner))
```
Lets also check for createdb privilege since that is what we do during create db.

anju15bharti replied:
`has_prive_of_role` makes sure that. So need not to check for createdb priv.

tanscorpio7 replied:
My point was this should look same as the check we do in create database plus the OR with login_is_db_owner.

HarshLunagariya replied:
We don't have a strong reason to/not rely on createdb privileges. My suggestion is to stick with existing check without createdb priv check.

Commit:
Merge remote-tracking branch 'origin/server-database-roles-4-x' into jira-dbcreator

Commit:
Fix Failure

tanscorpio7 commented on test/JDBC/expected/dbcreator_role-vu-verify.out:
```diff
@@ -0,0 +1,1257 @@
+-- tsql
```
Couple of negative cases we can test.
1. db_creator creates database -> transfer ownership to someone else -> cannot do any action insdie database i.e. does not map to dbo (but can still drop /alter database).
2. Create databases using dbcreator and sysadmin. remove the login from dbcreator and then check if restricted actions. alter/drop on existing dbs.

anju15bharti replied:
1. when we transfer the ownership through alter authorization, ownership gets revoked from previous login which will be the same case there and hence it will become normal case where it will be able to drop database but not alter until it has mapped user for that database. will add testcases as well.
2. We already have check that ensure that if login is member of both dbcreator and sysadmin, removed login from dbcreator it will still have attribute priv through sysadmin membership (ref: test/JDBC/expected/dbcreator_role-vu-verify.out:341)

Commit:
Added some testcases

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

HarshLunagariya commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6964,3 +6936,78 @@ bbf_set_tran_isolation(char *new_isolation_level_str)
+	revoke_createrole = (prev_createrole == 1 && createrole == 0);
+	revoke_createdb = (prev_createdb == 1 && createdb == 0);
+
+	// /* If both attributes need to be granted/revoked */
+	if ((grant_createrole && grant_createdb)
+		|| (revoke_createrole && revoke_createdb)) 
+		appendStringInfo(query, "ALTER ROLE dummy WITH %s %s; ", (grant_createrole) ? "createdb" :
+						 "nocreatedb", (grant_createrole) ? "createrole" : "nocreaterole");
+	
+	/* If only one attribute needs to be granted/revoked */
+	else if (grant_createrole || revoke_createrole
+		|| grant_createdb || revoke_createdb) 
+		appendStringInfo(query, "ALTER ROLE dummy WITH %s; ", (grant_createdb) ? "createdb" : 
+				((grant_createrole) ? "createrole" : ((revoke_createrole) ? "nocreaterole" : "nocreatedb")));
```
One more suggestion:
```suggestion
	appendStringInfo(query, "ALTER ROLE dummy WITH %s %s; ", grant_createrole ? "createrole" : (revoke_createrole ? "nocreaterole" : ""),
															 grant_createdb ? "createdb" : (revoke_createdb ? "nocreatedb" : ""));
```

HarshLunagariya commented on test/JDBC/expected/dbcreator_role-vu-verify.out:
```diff
@@ -0,0 +1,1338 @@
+-- terminate-tsql-conn user=dbcreator_login1 password=123
+
+
+-- tsql database=dbcreator_db1
+-- dbcreator login should not get mapped to dbo if no user exist, it should disconnect
+drop user dbcreator_user1
+go
+
+-- terminate-tsql-conn database=dbcreator_db1
+
+-- tsql user=dbcreator_login1 password=123
+-- it should disconnect
+use dbcreator_db1
+go
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: The server principal "dbcreator_login1" is not able to access the database "dbcreator_db1" under the current security context)~~
```
Can you please explain this case? What are we trying to test?

anju15bharti replied:
It's a negative case that if dbcreator login don't have mapped user in the database, it should disconnect and should not get mapped to dbo by any chance.

HarshLunagariya replied:
Alright, Got it.

Commit:
Enhance query creation condition

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Terminate psql connection in test_windows_login-vu-prepare testfile

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

Commit:
Increase SLA in test_windows_login-vu-prepare testfile

Signed-off-by: ANJU BHARTI <abanju@amazon.com>

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6964,3 +6936,71 @@ bbf_set_tran_isolation(char *new_isolation_level_str)
+	if (has_privs_of_role(login_oid, get_securityadmin_oid()))
+	{
+		createrole++;
+		is_securityadmin = 2;
```
lets use true and false for bools

Commit:
Use has_privs_of_role instead is_role_member

Signed-off-by: ANJU BHARTI <abanju@amazon.com>
